disnake/enums.py

```diff
@@ -72,326 +85,1370 @@
+        self.member_map[name] = value
+        self.value_map[value] = name
 
 
 class EnumMeta(type):
-    if TYPE_CHECKING:
-        __name__: ClassVar[str]
-        _enum_member_names_: ClassVar[List[str]]
-        _enum_member_map_: ClassVar[Dict[str, Any]]
-        _enum_value_map_: ClassVar[Dict[Any, Any]]
-
-    def __new__(cls, name, bases, attrs, *, comparable: bool = False):
-        value_mapping = {}
-        member_mapping = {}
-        member_names = []
-
-        value_cls = _create_value_cls(name, comparable)
-        for key, value in list(attrs.items()):
-            is_descriptor = _is_descriptor(value)
-            if key[0] == "_" and not is_descriptor:
-                continue
-
-            # Special case classmethod to just pass through
-            if isinstance(value, classmethod):
-                continue
-
-            if is_descriptor:
-                setattr(value_cls, key, value)
-                del attrs[key]
-                continue
-
-            try:
-                new_value = value_mapping[value]
-            except KeyError:
-                new_value = value_cls(name=key, value=value)
-                value_mapping[value] = new_value
-                member_names.append(key)
-
-            member_mapping[key] = new_value
-            attrs[key] = new_value
-
-        attrs["_enum_value_map_"] = value_mapping
-        attrs["_enum_member_map_"] = member_mapping
-        attrs["_enum_member_names_"] = member_names
-        attrs["_enum_value_cls_"] = value_cls
-        actual_cls = super().__new__(cls, name, bases, attrs)
-        value_cls._actual_enum_cls_ = actual_cls  # type: ignore
-        return actual_cls
-
-    def __iter__(cls):
-        return (cls._enum_member_map_[name] for name in cls._enum_member_names_)
-
-    def __reversed__(cls):
-        return (cls._enum_member_map_[name] for name in reversed(cls._enum_member_names_))
-
-    def __len__(cls):
-        return len(cls._enum_member_names_)
-
-    def __repr__(cls):
+    __is_enum_instantiated: ClassVar[bool] = False
+
+    _name_map_: ClassVar[Mapping[str, Enum]]
+    _value2member_map_: ClassVar[Mapping[Any, Enum]]
+
+    def __new__(
+        metacls: Type[_T],  # pyright: reportSelfClsParameterName=false
+        name: str,
+        bases: Tuple[Type[Any], Type[Any]],
+        namespace: _EnumDict,
+    ) -> _T:
+
+        if not EnumMeta.__is_enum_instantiated:
+            EnumMeta.__is_enum_instantiated = True
+            return super().__new__(metacls, name, bases, namespace)
+
+        base, enum_type = bases  # ensured possible in __prepare__
+
+        ns: Dict[str, Any] = {
+            "__objtype__": base,
+            "__enumtype__": enum_type,
+            "_name_map_": (name_map := {}),
+            "_value2member_map_": (value_map := {}),
+            **{
+                name_: value_
+                for name_, value_ in Enum.__dict__.items()
+                if name_ not in ("__class__", "__module__", "__doc__")
```

Comment:
Why these three in particular? Would be great to have a comment explaining parts of this

Reply:
These can actually be removed now, after some testing, as they're overwritten by the new namespace anyways. In an earlier implementation this wasn't the case, as I was updating the other way around (a selection of old stuff over new stuff, instead of the current implementation's new stuff over old stuff).

That also explains why this was necessary-- without this extra check, the class, module and docstring fields of any new enum would be overwritten with that of the base enum class.

I'll do a bit of further testing to confirm, but I'm fairly certain this can be removed altogether.

Reply:
Yup, this check doesn't seem to be of any importance anymore due to the reasons explained in my previous comment. This code has been removed as part of 662e0bdf1f0c47a3142d4214b1cbda16fb6e5b17. We now copy the entire namespace over, as these bits get overwritten by the new namespace anyways.